### PR TITLE
Issue#79 jwt validation after login closes #79

### LINKED ISSUE #79: Add Middleware Validation using JWT for routes
Currently, the Login POST route creates a JWT Token for authorising the login. 
This should be used to validate if a user is logged in before allowing to add listings, etc.

### PR DESCRIPTION
**Pull request by RonnieDsouza**
**Description** 
After a user has been logged in, they receive a JWT Token. 
This token must be used to authenticate the user when they try to access certain routes.

**Changelog**
- Added ```validateLoggedIn``` Middleware
  - Checks for the JWT Token in the Header (Authorization).
- Added said Middleware to POST ```addListing``` route. 

**Screenshots**
The user receives a token upon successful login.
image

That token is included in the Authorization Header when accessing certain routes. 
(Currently only the POST ```addListing``` route.)
image

When a token is invalid (or missing). The authorization fails and access is denied. 
- Removal of first letter of token -> Making it invalid
image
 
- Missing token
image

### CODE REVIEW

File: server/error/api-error.js [hunk 1]
```diff
@@ -8,6 +8,10 @@ class ApiError {
     return new ApiError(400, msg);
   }
 
+  static unAuthenticated(msg) {
```
Comment (jcsnorlax97) [inline server/error/api-error.js:11]: I'm just being picky. Isn't "unauthenticated" a single word? lol

File: server/middleware/validateLoggedIn.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+const validateLoggedIn = (req, res, next) => {
+  try {
+    const token = req.headers.authorization.split(' ')[1];
+    const decoded = jwt.verify(token, process.env.JWT_KEY);
+    req.userData = decoded;
+    next();
+  } catch (err) {
+    // Invalid auth token, meaning not logged in
+    next(ApiError.unAuthenticated('You need to be logged in to do this.'));
```
Comment (jcsnorlax97) [inline server/middleware/validateLoggedIn.js:12]: unAuthenticated vs unauthenticated lol

**Review by jcsnorlax97 — APPROVED** [review]
Looks good to me in general. Approved!

**Review by AzizulHakim96 — APPROVED** [review]

**Review by Mik3Qiao — APPROVED** [review]

**Review by NathanCarriere — APPROVED** [review]

File: server/error/api-error.js (hunk 1 above)
Comment (RonnieDsouza) [reply server/error/api-error.js:11]: Yeah! I was thinking about that too! But I liked the look of it this way lol, kinda easier to read. But either way is oK!

File: server/middleware/validateLoggedIn.js (hunk 2 above)
Comment (RonnieDsouza) [reply server/middleware/validateLoggedIn.js:12]: yeah 😅

**Pull request merged**